src/alexandria3k/__main__.py

```diff
@@ -99,10 +99,82 @@ def get_data_source_instance(args):
     return class_(data_location, sample, args.attach_databases)
 
 
+def download(args):
+    """Download data using the specified data source."""
+    args.validate_args(args)
+    data_source_instance = get_data_source_instance(args)
+    if not hasattr(data_source_instance, "download"):
+        raise Alexandria3kError(
+            f"The data source {args.data_name} does not support downloading"
+        )
+    data_source_instance.download(
+        args.database, args.sql_query, args.output_path, *args.extra_args
+    )
+    perf.log(
+        f"Data downloaded and saved to {data_source_instance.output_path}"
+    )
+
+
+def validate_args(args):
+    """Validate that both database and sql_query are either both provided or both omitted."""
+    if (args.database and not args.sql_query) or (
+        args.sql_query and not args.database
+    ):
+        raise argparse.ArgumentTypeError(
+            "Both --database and --sql-query must be provided together or not at all."
+        )
+    return args
+
+
+def add_subcommand_download(subparsers):
+    """Add the arguments of the download subcommand."""
+    parser = subparsers.add_parser(
+        "download", help="Download data using the specified data source."
+    )
+    parser.set_defaults(func=download)
+    parser.add_argument(
+        "database", help="File path of the database to use", nargs="?"
+    )
+    parser.add_argument(
+        "data_name",
+        choices=facility_names("data_sources"),
+        help="Name of the data source to use",
+    )
+    parser.add_argument(
+        "-s",
+        "--sql-query",
+        type=str,
+        help="SQL query to retrieve the data for downloading",
+    )
+    parser.add_argument(
+        "-o",
+        "--output-path",
+        type=str,
+        nargs="?",
+        help="File path to save the downloaded data",
+    )
+    parser.add_argument(
+        "extra_args",
+        nargs="*",
+        help="Additional arguments for the data source (e.g. URL, key, file path)",
+    )
+
+    # Add a custom validation function to the parser
+    parser.set_defaults(validate_args=validate_args)
+
+
 def populate(args):
     """Populate the specified database from the specified data source."""
 
     data_source_instance = get_data_source_instance(args)
+    if hasattr(data_source_instance, "download"):
+        # Check if the ouput_path attribute is not None
+        if data_source_instance.output_path is None:
+            raise Alexandria3kError(
```

Comment:
Please also take into account the CLI use. There the program is invoked twice: to download and then to populate.

Reply:
I am referring to the [CLI](https://dspinellis.github.io/alexandria3k/cli-eg.html), not the [API](https://dspinellis.github.io/alexandria3k/python-eg.html).

Reply:
Hmmm, you got a point, should we do separate checks for the API (like the one we have) and the CLI , or should we do it with an another way? What do you suggest?

Reply:
I am a little confused, should we check if both are called in the same time? For the `download` we give the database path to create the tables into and for the `populate` function to just populate in the same database. At least in our case, it isn't mandatory to first call the `download` function, it just makes more sense.